third_party/com.google.cloud.tools.eclipse.jst.server.core/src/com/google/cloud/tools/eclipse/jst/server/core/BasePublishOperation.java

```diff
@@ -0,0 +1,359 @@
+   */
+  protected abstract IPath getRuntimeBaseDirectory();
+
+  /**
+   * Return the location to where this server expects to find the given module. This location is
+   * typically under {@link #getRuntimeBaseDirectory()}.
+   */
+  protected abstract IPath getModuleDeployDirectory(IModule module);
+
+
+  /**
+   * Update the server's notion of the module's publish state
+   * 
+   * @see IServer#PUBLISH_STATE_NONE
+   * @see IServer#PUBLISH_STATE_UNKNOWN
+   * @see IServer#PUBLISH_STATE_INCREMENTAL
+   * @see IServer#PUBLISH_STATE_FULL
+   */
+  protected abstract void setModulePublishState(IModule[] module, int publishState);
+
+  /**
+   * Determine the changed resources since last published.
+   */
+  protected abstract IModuleResourceDelta[] getPublishedResourceDelta(IModule[] module);
+
+  /**
+   * Return all the resources to be published for the given module.
+   */
+  protected abstract IModuleResource[] getResources(IModule[] module);
+
+  /**
+   * Return {@code true} if this server can serve modules without explicitly publishing.
+   */
+  protected abstract boolean isServeModulesWithoutPublish();
+
+  public int getOrder() {
+    return 0;
+  }
+
+  public int getKind() {
+    return REQUIRED;
+  }
+
+  /**
+   * Initializes this instance.
+   */
+  protected void initialize() throws CoreException {
+    IPath base = getRuntimeBaseDirectory();
+    if (base != null) {
+      helper = new PublishHelper(base.append("temp").toFile());
+    } else {
+      // We are doomed without a base directory. However, allow the catastrophe
+      // to occur elsewhere and hope for a useful error message.
+      helper = new PublishHelper(null);
+    }
+  }
+
+
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    initialize();
+
+    List<IStatus> status = new ArrayList<IStatus>();
+    // If parent web module
+    if (module.length == 1) {
+      publishDir(module[0], status, monitor);
+    } else {
+      // Else a child module
+      Properties p = loadModulePublishLocations();
+
+      // Try to determine the URI for the child module
+      IWebModule webModule = (IWebModule) module[0].loadAdapter(IWebModule.class, monitor);
+      String childURI = null;
+      if (webModule != null) {
+        childURI = webModule.getURI(module[1]);
+      }
+      // Try to determine if child is binary
+      IJ2EEModule childModule = (IJ2EEModule) module[1].loadAdapter(IJ2EEModule.class, monitor);
+      boolean isBinary = false;
+      if (childModule != null) {
+        isBinary = childModule.isBinary();
+      }
+
+      if (isBinary) {
+        publishArchiveModule(childURI, p, status, monitor);
+      } else {
+        publishJar(childURI, p, status, monitor);
+      }
+      saveModulePublishLocations(p);
+    }
+    throwException(status);
+    setModulePublishState(module, IServer.PUBLISH_STATE_NONE);
+  }
+
+  private void publishDir(IModule module2, List<IStatus> status, IProgressMonitor monitor)
```

Comment:
suggest status --> statuses (plural) and stat --> to status below. Otherwise, stat sounds like a statistic, not a status